### PR TITLE
Fixes #1873 Implemented remaining 7/8 of ImageDrawLine

### LINKED ISSUE #1873: [textures] ImageDrawLine only covers one octant
### Issue description

ImageDrawLine only covers the octant where `dx > dy && dx > 0 && dy > 0`.

### Environment

Win32 (issue is platform agnostic)

### Issue Screenshot

graph

### Code Example

This code only draws three of the 16 lines tested for.
 (tests 1 for each octant + 1 for each dividing line)
```
#include "raylib.h"
const int size = 121, mid = 60, quart = 30;
void DrawFromMidToHere(Image* image, int x, int y){
    ImageDrawLine(image, mid, mid, x, y, BLACK); 
}
int main(){
    Image im = GenImageColor(size, size, WHITE);
    DrawFromMidToHere(&im,size-1,size-1);
    DrawFromMidToHere(&im,size-1,0);
    DrawFromMidToHere(&im,0,0);
    DrawFromMidToHere(&im,0,size-1);
    for(int i = 0; i<2; i++){
        DrawFromMidToHere(&im,i*size, mid - quart);
        DrawFromMidToHere(&im,i*size, mid);
        DrawFromMidToHere(&im,i*size, mid + quart);
    }
    for(int i = 0; i<2; i++){
        DrawFromMidToHere(&im,mid - quart, i*size);
        DrawFromMidToHere(&im,mid,         i*size);
        DrawFromMidToHere(&im,mid + quart, i*size);
    }
    ExportImage(im, "./graph.png");
    return 0;
}
```

### PR DESCRIPTION
**Pull request by Buhlean**
Fixes #1873
The existing code was correct for one octant, it now works for all 8
Added two internal functions, _ImageDrawLineHorizontal and _ImageDrawLineVertical, without which it would've been 4 times as much code.
graph

### CODE REVIEW

**Comment by Buhlean** [comment]
That was 6 hours! :D 
I refused to look at anyone's code, so it's my fault :) 

Here's a quick test snippet, it produced the above image:
```
#include "raylib.h"
const int size = 121, mid = 60, quart = 30;
void DrawFromMidToHere(Image* image, int x, int y){
    ImageDrawLineQ(image, mid, mid, x, y, BLACK); 
}
int main(){
    Image im = GenImageColor(size, size, WHITE);
    DrawFromMidToHere(&im,size-1,size-1);
    DrawFromMidToHere(&im,size-1,0);
    DrawFromMidToHere(&im,0,0);
    DrawFromMidToHere(&im,0,size-1);
    for(int i = 0; i<2; i++){
        DrawFromMidToHere(&im,i*size, mid - quart);
        DrawFromMidToHere(&im,i*size, mid);
        DrawFromMidToHere(&im,i*size, mid + quart);
    }
    for(int i = 0; i<2; i++){
        DrawFromMidToHere(&im,mid - quart, i*size);
        DrawFromMidToHere(&im,mid,         i*size);
        DrawFromMidToHere(&im,mid + quart, i*size);
    }
    ExportImage(im, "./graph.png");
    return 0;
}
```

**Comment by raysan5** [comment]
@Buhlean I always try to avoid internal functions on shapes drawing. I feel this problem could be simplified in some other way...

**Comment by Buhlean** [comment]
I crammed it all into one function.
Taking away a core tool of algorithm design (functions) will make some algorithms hideous and a copy-paste + search-and-replace nightmare.
It works, it only calculates what it needs to at every stage, it is compiler friendly (as in the compiler will know what i mean and can produce decent instructions). Maintenance will be worse, though.

**Comment by Buhlean** [comment]
Wait, i figured out a better way. Ignore the changes for now!

**Comment by Buhlean** [comment]
Thanks for pushing me, it's so much better now!
I substituted sX, sY, eX, eY with abstract "function parameters" U and V, then wrote one loop using those.
Maths! 
Also, lots of comments because it's hairy enough already.

File: src/textures.c [hunk 1]
```diff
@@ -2414,26 +2414,108 @@ void ImageDrawPixelV(Image *dst, Vector2 position, Color color)
...
-
+     
 // Draw line within an image
-void ImageDrawLine(Image *dst, int startPosX, int startPosY, int endPosX, int endPosY, Color color)
-{
-    int m = 2*(endPosY - startPosY);
-    int slopeError = m - (endPosX - startPosX);
-
-    for (int x = startPosX, y = startPosY; x <= endPosX; x++)
+void ImageDrawLine (Image *dst, int startPosX, int startPosY, int endPosX, int endPosY, Color color){
```
Comment (raysan5) [inline src/textures.c:2419]: Please, code you review raylib coding conventions?

File: src/textures.c [hunk 2]
```diff
@@ -2414,26 +2414,108 @@ void ImageDrawPixelV(Image *dst, Vector2 position, Color color)
...
+    // Using Bresenham's algorithm as described in 
+    // Drawing Lines with Pixels - Joshua Scott - March 2012
+    // https://classic.csunplugged.org/wp-content/uploads/2014/12/Lines.pdf
+    
+    int     changeInX = (endPosX - startPosX);
+    int abs_changeInX = changeInX < 0 ? -changeInX : changeInX;
```
Comment (raysan5) [inline src/textures.c:2425]: Please, code you review raylib coding conventions?

File: src/textures.c [hunk 3]
```diff
@@ -2414,26 +2414,108 @@ void ImageDrawPixelV(Image *dst, Vector2 position, Color color)
...
+
+    int is_x_y_reversed = (abs_changeInY < abs_changeInX);
+    
+    if (is_x_y_reversed) 
     {
-        ImageDrawPixel(dst, x, y, color);
-        slopeError += m;
-
-        if (slopeError >= 0)
+        A =     2* abs_changeInY;
```
Comment (raysan5) [inline src/textures.c:2439]: Please, code you review raylib coding conventions?

File: src/textures.c [hunk 4]
```diff
@@ -2414,26 +2414,108 @@ void ImageDrawPixelV(Image *dst, Vector2 position, Color color)
...
+            startU = startPosX;
+            startV = startPosY;
+            endU   =   endPosX;
+          //endV   =   endPosY;
+        }
+        else
         {
-            y++;
-            slopeError -= 2*(endPosX - startPosX);
+            startU =   endPosX;
```
Comment (raysan5) [inline src/textures.c:2452]: Please, could you avoid vertical alignment of variables?

File: src/textures.c [hunk 5]
```diff
@@ -2414,26 +2414,108 @@ void ImageDrawPixelV(Image *dst, Vector2 position, Color color)
...
+            startV =   endPosY;
+            endU   = startPosX;
+          //endV   = startPosY;
+            
+            // since start and end are reversed:
+            changeInX = -changeInX;
+            changeInY = -changeInY;
+        }
+        
+        V_step = changeInY < 0 ? -1 : 1;
```
Comment (raysan5) [inline src/textures.c:2462]: Please, code you review raylib coding conventions?

**Comment by raysan5** [comment]
@Buhlean Please, code you review raylib coding conventions?

**Comment by raysan5** [comment]
@Buhlean thanks for the improvement!

**Pull request merged**